### ISSUE TITLE
Infinite recursion in transpiler.coffee

### ISSUE THREAD

**Issue by Daniel15** [gandm/language-babel#66]
This call appears to recurse infinitely when using Nuclide to edit files on a remote server:
```
return @getLocalConfig path.dirname(fromDir), toDir, localConfig
```

This exception is thrown on save:
```
RangeError: Maximum call stack size exceeded
  at RegExp.exec (native)
  at posixSplitPath (path.js:385:22)
  at Object.posix.dirname (path.js:513:16)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:35)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  at Transpiler.getLocalConfig (/Users/danlo/.atom/packages/language-babel/lib/transpiler.coffee:174:15)
  ... and so on...
```

I have "Transpile on save" turned off.

**Comment by Daniel15** [comment]
In the meantime, I downgraded to the previous version by running this command:

```
apm install language-babel@0.13.6
```

**Comment by gandm** [comment]
Oh! It works ok on my native filesystem. It also seems to run TravisCI tests for OSX ok. I don't have Nuclide setup so I can't test it on a remote fs. I'm guessing it's something to do with the the compare on line 174 below in transpiler.coffee. I've put in a console.log line at 173 to see what is happening as it traverses paths to read any possible local config files. Even with transpile on save off I still read these files to look for the transpileOnSave property values that may be local to a directory - that's why turning off the global value made no difference. If you can add something like the code in 173 it would help if not then I'll put together some debug code and publish it. 

![image](https://cloud.githubusercontent.com/assets/2313237/9863204/d8edff94-5b35-11e5-971f-6bd0b354d122.png)

**Comment by Daniel15** [comment]
This is what the output looks like:
```
from=nuclide://myserver.example.com:8084/data/users/danlo/www-hg/html/js/modules/utils to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data/users/danlo/www-hg/html/js/modules to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data/users/danlo/www-hg/html/js to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data/users/danlo/www-hg/html to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data/users/danlo/www-hg to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data/users/danlo to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data/users to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084/data to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide://myserver.example.com:8084 to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=nuclide:/ to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
from=. to=nuclide:/myserver.example.com:8084/data/users/danlo/www-hg
```

Maybe a temporary hack could be to avoid doing the search if the path contains `nuclide://`? @bolinfest (who works on Nuclide) probably has more insights here.

**Comment by bolinfest** [comment]
https://github.com/gandm/language-babel/pull/67 is my proposed fix to this issue. I don't want others to have to special-case `nuclide://`.

**Comment by gandm** [comment]
I've just pushed some changes to the repo that should fix your problem. I won't publish until you've checked it works . I've added another config option that is disabled by default `allowLocalOverride` that will stop traversal of the directory. I also trap for infinite recursion should the setting be on.

**Comment by gandm** [comment]
Published 0.15.0 that should fix the issue.

**Comment by Daniel15** [comment]
Yes, I just confirmed that 0.15.0 fixes the issue. Thank you for being so responsive, it's really appreciated :smile:
